File: trials/haskell-simple-ebnf/src/EBNF.hs
module EBNF
    ( TStatement(..)
    , statements
    ) where

import Text.Parsec as P

data TOperator =
    TConcatenation
  | TAlternation
    deriving (Show)

data TExpression =
    TTerminal String
  | TRule String
  | TBinary TExpression TOperator TExpression
    deriving (Show)

data TStatement =
    TDefinition TExpression TExpression
  | TStatements [TStatement]
      deriving (Show)

rule :: P.Parsec String st TExpression
rule = TRule <$> (P.try $ P.char '<' *> ruleDef <* P.char '>' <|> ruleDef)
  where
    ruleDef = (\a b -> a:b) <$> P.letter <*> (P.many alpha)
    alpha = P.letter <|> P.digit

operator :: P.Parsec String st TOperator
operator = choose <$> (P.oneOf ",|")
  where
    choose ',' = TConcatenation
    choose '|' = TAlternation

expression :: P.Parsec String st TExpression
expression = P.try binaryOp <|> part
  where
    binaryOp = TBinary <$> part <*> operator <*> expression
    part = TTerminal <$> (P.char '\'' *> (P.many1 $ P.noneOf "\'") <* P.char '\'')
      <|> rule

statements :: P.Parsec String st TStatement
statements = TStatements <$> P.many1 definition
  where
    definition = TDefinition <$> (rule <* P.char '=') <*> (expression <* P.char ';')

File: trials/fsharp-simple-ebnf/fsharp-simple-ebnf/EBNF.fs
module EBNF

open FParsec

// AST
// ---
type TOperator =
      TConcatenation
    | TAlternation

type TExpression =
      TTerminal of string
    | TRule of string
    | TBinary of TExpression * TOperator * TExpression

type TStatement =
      TDefinition of TExpression * TExpression
    | TStatements of TStatement list

// Parsers
// -------
let rule: Parser<TExpression, unit> =
    let alpha = letter <|> digit
    let ruleDef = many1Chars2 letter alpha
    ((pchar '<' >>. ruleDef .>> pchar '>') <|> ruleDef) |>> TRule

let operator: Parser<TOperator, unit> =
    let choose c =
        match c with
        | ',' -> TConcatenation
        | '|' -> TAlternation
    anyOf ",|" |>> choose

let expression, expressionRef = createParserForwardedToRef<TExpression, unit> ()

let statements: Parser<TStatement, unit> =
    let definition = pipe2 (rule .>> pchar '=') (expression .>> pchar ';') (fun a b -> TDefinition (a, b))
    many1 definition |>> TStatements

expressionRef :=
    let terminal = (pchar '\'' >>. many1Chars (noneOf "'") .>> pchar '\'') |>> TTerminal
    let part = terminal <|> rule
    let binaryOp = pipe3 part operator expression (fun a b c -> TBinary (a, b, c))
    attempt binaryOp <|> part
